Sources/Basics/Environment/Environment.swift

```diff
@@ -0,0 +1,326 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+#if USE_IMPL_ONLY_IMPORTS
+@_implementationOnly import TSCclibc
+#else
+import TSCclibc
+#endif
+
+// FIXME: Use Synchronization.Mutex when available
+class Mutex<T>: @unchecked Sendable {
+    var lock: NSLock
+    var value: T
+
+    init(value: T) {
+        self.lock = .init()
+        self.value = value
+    }
+
+    func withLock<U>(_ body: (inout T) -> U) -> U {
+        self.lock.lock()
+        defer { self.lock.unlock() }
+        return body(&self.value)
+    }
+}
+
+// FIXME: This should come from Foundation
+// FIXME: package (public required by users)
+public struct Environment {
+    var storage: [EnvironmentKey: String]
+}
+
+// MARK: - Accessors
+extension Environment {
+    package init() {
+        self.storage = .init()
+    }
+
+    package subscript(_ key: EnvironmentKey) -> String? {
+        _read { yield self.storage[key] }
+        _modify { yield &self.storage[key] }
+    }
+}
+
+// MARK: - Conversions between Dictionary<String, String>
+extension Environment {
+    package init(_ dictionary: [String: String]) {
+        self.storage = .init()
+        let sorted = dictionary.sorted { $0.key < $1.key }
+        for (key, value) in sorted {
+            self.storage[.init(key)] = value
+        }
+    }
+}
+
+extension Dictionary<String, String> {
+    package init(_ environment: Environment) {
+        self.init()
+        let sorted = environment.sorted { $0.key < $1.key }
+        for (key, value) in sorted {
+            self[key.rawValue] = value
+        }
+    }
+}
+
+// MARK: - Path Modification
+extension Environment {
+    package mutating func prependPath(key: EnvironmentKey, value: String) {
+        guard !value.isEmpty else { return }
+        if let existing = self[key] {
+            self[key] = "\(value)\(Self.pathValueDelimiter)\(existing)"
+        } else {
+            self[key] = value
+        }
+    }
+
+    package mutating func appendPath(key: EnvironmentKey, value: String) {
+        guard !value.isEmpty else { return }
+        if let existing = self[key] {
+            self[key] = "\(existing)\(Self.pathValueDelimiter)\(value)"
+        } else {
+            self[key] = value
+        }
+    }
+
+    package static var pathValueDelimiter: String {
```

Comment:
I think `pathEntryDelimiter` is better - the variable is a set of key, value pairs which makes value a bit overloaded.

Reply:
Yeah, I can change the name to your suggestion